pkg/kubelet/kubelet_pods.go

```diff
@@ -1673,11 +1674,11 @@ func getPhase(pod *v1.Pod, info []v1.ContainerStatus, podIsTerminal bool) v1.Pod
 	}
 
 	switch {
-	case pendingInitialization > 0 &&
+	case pendingRegularInitContainers > 0 || (pendingRestartableInitContainers > 0 &&
 		// This is needed to handle the case where the pod has been initialized but
 		// the restartable init containers are restarting and the pod should not be
 		// placed back into v1.PodPending since the regular containers have run.
-		!kubecontainer.HasAnyRegularContainerStarted(&spec, info):
+		!kubecontainer.HasAnyRegularContainerStarted(&spec, info)):
```

Comment:
or maybe we have to change this function?

to only return true if there is any regular container that is created or running but not terminated?

Reply:
> or maybe we have to change this function?
> 
> to only return true if there is any regular container that is created or running but not terminated? (as the regular container may be in the terminated state after the node reboot while it hasn't started before)

https://github.com/kubernetes/kubernetes/blob/746f08a8da0a45e8b1a46362822db05cb11ed294/pkg/kubelet/container/helpers.go#L410-L412

Currently, relying solely on the function HasAnyRegularContainerStarted is indeed not sufficient，when `status.State.Running != nil || status.State.Terminated != nil` occurs and `HasAnyRegularContainerStarted `returns true, this is reasonable when the node has not been restarted. So do we have a way to determine whether the node has restarted?

Reply:
I am ok with this change. This will correctly differentiate the initialization of init containers.

However, at the moment, I don't have a good way to differentiate the restarting restartable init containers from the initialization phase.

I'd prefer false Pending than false Running so I suggest you to change the `HasAnyRegularContainerStarted` to only return true when the regular container is in Running state.

If you are hesitating to change it, it's good to merge this as it is. I'll try to complement this by changing HasAnyRegularContainerStarted with the follow-up PR.

Reply:
> I'd prefer false Pending than false Running so I suggest you to change the HasAnyRegularContainerStarted to only return true when the regular container is in Running state.

I prefer doing it this way currently, because, it is better for  unexpected  pod phase to occur during node restart than when the node is running normally.

>If you are hesitating to change it, it's good to merge this as it is. I'll try to complement this by changing HasAnyRegularContainerStarted with the follow-up PR.

ok，I will also pay attention to future improvements..